src/fabric.c

```diff
@@ -602,17 +602,30 @@ int DEFAULT_SYMVER_PRE(fi_getinfo)(uint32_t version, const char *node,
 		if (FI_VERSION_LT(prov->provider->fi_version, version))
 			continue;
 
-		if (ofi_is_util_prov(prov->provider) &&
-		    (flags & OFI_CORE_PROV_ONLY))
-			continue;
-
-		if (util_len && util_name) {
-			if (strncasecmp(util_name, prov->provider->name, util_len))
-				continue;
-		} else if (core_len && core_name) {
-			if (!ofi_is_util_prov(prov->provider) &&
-			    strncasecmp(core_name, prov->provider->name, core_len))
+		if (ofi_is_util_prov(prov->provider)) {
+			/* This is util provider */
+			if ((util_len && util_name) &&
+			    !(flags & OFI_CORE_PROV_ONLY)) {
+				/* If util name consists in provider name
+				 * and no OFI_CORE_PROV_ONLY flag is set,
+				 * FI info of util provider is requested */
+				if (strncasecmp(util_name,
+						prov->provider->name,
+						util_len))
+					continue;
+			} else {
+				/* Requested core provider's FI info 
+				 * or provider name doesn't contain util name */
 				continue;
+			}
+		} else {
+			/* This is core provider */
+			if (core_len && core_name) {
+				if (strncasecmp(core_name,
```

Comment:
Ok, thank you for your comment. But I'm not sure that I got your comment correctly. Please, check my understanding of that.